R/deploy-site.R

```diff
@@ -0,0 +1,91 @@
+#' Deploy a pkgdown site on Travis-CI to Github Pages
+#'
+#' `deploy_site_github()` sets up your SSH keys for deployment, builds the
+#' site with [build_site()], commits the site to the `gh-pages` branch and then pushes
+#' the results back to GitHub. `deploy_site_github()` is meant only to be used
+#' by the CI system on Travis, it should not be called locally. See 'Setup' for
+#' details on setting up your repository to use this.
+#'
+#' @section Setup:
+#' Add the following to your `.travis.yml` file.
+#'
+#' ```
+#' before_deploy: Rscript -e 'install.packages("pkgdown")'
+#' deploy:
+#'   provider: script
+#'   script: Rscript -e 'pkgdown::deploy_site_github()'
+#' ```
+#'
+#' Then you will need to setup your deployment keys. The easiest way is to call
+#' `travis::use_travis_deploy()`. This will generate and push the necessary
+#' keys to your GitHub and Travis accounts. See the [travis package
+#' website](https://ropenscilabs.github.io/travis/) for more details.
+#'
+#' @inheritParams build_site
+#' @param tarball The location of the built package tarball. The default Travis
+#'   configuration for R packages sets `PKG_TARBALL` to this path.
+#' @param ssh_id The private id to use, a base64 encoded content of the private
+#'   pem file. This should _not_ be your personal private key. Instead create a
+#'   new keypair specifically for deploying the site. The easiest way is to use
+#'   `travis::use_travis_deploy()`.
+#' @param repo_slug The `user/repo` slug for the repository.
+#' @param commit_message The commit message to be used for the commit.
+#' @param verbose Print verbose output
+#' @param ... Additional arguments passed to [build_site()].
+#' @export
+deploy_site_github <- function(
+  pkg = ".",
+  tarball = Sys.getenv("PKG_TARBALL", ""),
+  ssh_id = Sys.getenv("id_rsa", ""),
+  repo_slug = Sys.getenv("TRAVIS_REPO_SLUG", ""),
+  commit_message = construct_commit_message(pkg),
+  verbose = FALSE,
+  ...) {
+
+  if (!nzchar(tarball)) {
+    stop("No built tarball detected, please provide the location of one with `tarball`", call. = FALSE)
+  }
+
+  if (!nzchar(ssh_id)) {
+    stop("No deploy key found, please setup with `travis::use_travis_deploy()`", call. = FALSE)
+  }
+
+  if (!nzchar(repo_slug)) {
+    stop("No repo detected, please supply one with `repo_slug`", call. = FALSE)
+  }
+
+  rule("Deploying site", line = 2)
+  rule("Installing package", line = 1)
+  callr::rcmd("INSTALL", tarball, show = verbose, fail_on_status = TRUE)
+
+  build_site(pkg, ...)
```

Comment:
I think we need to clone before building the documentation for the new default of

```
development:
  mode: auto
```

Otherwise, building dev-mode docs will purge existing release docs.